### PR TITLE
Upstream a couple of PTryFrom instances 

### PR DESCRIPTION
**Pull request by MangoIV**
## Instances

- [x] PMaybeData
- [x] PCredential
- [x] PStakingCredential
- [x] PPubKeyHash
- [x] PValidatorHash
- [x] PTokenName 
- [x] PCurrencySymbol
- [x] PMap 
- [x] PValue
- [x] PPOSIXTime

## Tests

- [x] Succeeding property tests
- [x] Test `Extra.ByteString` utility functions

### CODE REVIEW

**Comment by emiflake** [comment]
Could we also get:
- `PTryFrom PData (PAsData b) => PTryFrom PData (PAsData (DerivePNewtype c b))`?
Here's my instance:
```haskell
instance PTryFrom PData (PAsData b) => PTryFrom PData (PAsData (DerivePNewtype c b)) where
  type
    PTryFromExcess PData (PAsData (DerivePNewtype c b)) =
      PTryFromExcess PData (PAsData b)
  ptryFrom' d k =
    ptryFrom' @_ @(PAsData b) d $ k . first punsafeCoerce
```
- `PTryFrom PData (PAsData PBool)`

File: plutarch-extra/Plutarch/Extra/TermCont.hs [hunk 1]
```diff
@@ -16,3 +20,11 @@ pletC = tcont . plet
 -- | Like `pmatch` but works in a `TermCont` monad
 pmatchC :: PMatch a => Term s a -> TermCont s (a s)
 pmatchC = tcont . pmatch
+
+-- | Like `ptryFrom` but works in a `TermCont` monad
+ptryFromC :: forall b a s. PTryFrom a b => Term s a -> TermCont s (Term s b, Reduce (PTryFromExcess a b s))
+ptryFromC = tcont . ptryFrom
+
+-- | Like `ptryFom'` but works in a `TermCont` monad
+ptryFromC' :: forall a b s. PTryFrom a b => Term s a -> TermCont s (Term s b, Reduce (PTryFromExcess a b s))
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/TermCont.hs:29]: Why add this?

**Comment by MangoIV** [comment]
We already have an instance for DerivePNewtype, do you mean something else?

File: plutarch-extra/Plutarch/Extra/TermCont.hs (hunk 1 above)
Comment (MangoIV) [reply plutarch-extra/Plutarch/Extra/TermCont.hs:29]: Just so we have 1:1 correspondence between the ptryFrom and respecting TermCont methods, if you want I can scratch that.

Comment (L-as) [reply plutarch-extra/Plutarch/Extra/TermCont.hs:29]: You never use `ptryFrom'`, thus neither `ptryFromC'`.

**Comment by L-as** [comment]
@MangoIV That instance doesn't work correctly because of the lack of the `Coercible` constraint, which is fixed in my PR.

File: plutarch-extra/Plutarch/Extra/TermCont.hs (hunk 1 above)
Comment (MangoIV) [reply plutarch-extra/Plutarch/Extra/TermCont.hs]: got it, removed.

**Comment by L-as** [comment]
We need to clean this up and merge it for 1.2

File: Plutarch/Api/V1/Value.hs [hunk 2]
```diff
@@ -48,3 +51,47 @@ deriving via
...
+      pisHexDigit = phoistAcyclic $
+        plam $ \chr ->
+          (chr #<= 57 #&& 48 #<= chr)
+            #|| (chr #<= 70 #&& 65 #<= chr)
+            #|| (chr #<= 102 #&& 97 #<= chr)
+
+deriving via
+  DerivePNewtype (PAsData PCurrencySymbol) (PAsData PByteString)
+  instance
+    PTryFrom PData (PAsData PCurrencySymbol)
```
Comment (maciej-bendkowski) [inline Plutarch/Api/V1/Value.hs:92]: It seems that  `CurrencySymbol` should be treated similarly to other hashes, i.e. have fixed-size length of 28 bytes?

**Comment by maciej-bendkowski** [comment]
How about the following:

```hs
instance PTryFrom PData (PAsData PTxId) where
    type PTryFromExcess PData (PAsData PTxId) = Flip Term PTxId
    ptryFrom' opq = runTermCont $ do
        constr <- tcont $ plet (pasConstr # opq)

        idx <- tcont $ plet (pfstBuiltin # constr)
        tcont $ \f ->
            pif
                (idx #== 0)
                (f ())
                (ptraceError "has to use the first constructor")

        dat <- tcont $ plet (psndBuiltin # constr)
        tcont $ \f ->
            pif
                (plength # dat #== 1)
                (f ())
                (ptraceError "has to have a single field")

        ( wrapped :: Term _ (PAsData PByteString)
            , unwrapped :: Term _ PByteString
            ) <-
            tcont $ ptryFrom @(PAsData PByteString) (phead # dat)

        tcont $ \f ->
            pif
                (plengthBS # unwrapped #== 28)
                (f ())
                (ptraceError "a TxId should be 28 bytes long")

        let fields :: Term _ (PDataRecord '["_0" ':= PByteString])
            fields = pdcons # wrapped # pdnil

            ptxId :: Term _ PTxId
            ptxId = pcon . PTxId $ fields

        pure (pdata ptxId, ptxId)

```

**Pull request merged**